### ISSUE TITLE
switch-to-configuration: line 3: use: command not found

### ISSUE THREAD

**Issue by kalbasit** [DBCDK/morph#64]
Hello,

I'm trying to use morph to deploy a Raspberry-pi 3b+ from my x86_64 system. I was able to replicate this issue also deploying an x86_64 system from my x86_64 system and I'm not sure why this is happening. Please find the error log, as well as all of the files used for the morph build.

Error log:

```console
 pure  λ  SSH_USER=root morph deploy nixos/testlab.nix boot
Selected 1/1 hosts (name filter:-0, limits:-0):
          0: 172.25.10.151 (secrets: 0, health checks: 0)

/nix/store/122d3ylj2b4c50a0aidv5l7ikasilrc6-morph
nix result path:
/nix/store/122d3ylj2b4c50a0aidv5l7ikasilrc6-morph

Pushing paths to 172.25.10.151:
        * /nix/store/wckw45dkz19ncfi5nzfb736qnhzv77zg-nixos-system-kore-19.09pre-git
        * /nix/store/l8b8np9l8ry2r9qga0l8m4zp9gyn2k92-nixos-system-kore-19.09pre-git.drv

Executing 'boot' on matched hosts:

** 172.25.10.151
/nix/store/wckw45dkz19ncfi5nzfb736qnhzv77zg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 3: use: command not found
/nix/store/wckw45dkz19ncfi5nzfb736qnhzv77zg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 4: use: command not found
/nix/store/wckw45dkz19ncfi5nzfb736qnhzv77zg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 5: use: command not found
/nix/store/wckw45dkz19ncfi5nzfb736qnhzv77zg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 6: use: command not found
/nix/store/wckw45dkz19ncfi5nzfb736qnhzv77zg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 7: use: command not found
/nix/store/wckw45dkz19ncfi5nzfb736qnhzv77zg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 8: syntax error near unexpected token `('
/nix/store/wckw45dkz19ncfi5nzfb736qnhzv77zg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 8: `use Sys::Syslog qw(:standard :macros);'
Error while activating new configuration.% 
```

Nixpkgs is pinned to https://github.com/NixOS/nixpkgs/commit/0419d983480a3dbc8583ac55c3dc849d2f65ccc2.

<details>
<summary>testlab.nix</summary>

```nix
let
  pkgs = import ../../../vendor/nix/nixpkgs.nix {};
in {
  network = {
    inherit pkgs;
    description = "KeepTruckin Nix builders";
  };

  # TODO: This does not work. I have to use the IP directly as the name of the
  # host instead of naming it kore and using deployment.targetHost. When I do
  # that, I get kore: No such file or directory.
  # "kore" = { config, pkgs, ... }: {
  #   imports = [
  #     ./hosts/kore
  #   ];
  #
  #   deployment.targetHost = "172.25.10.151";
  # };

  "172.25.10.151" = import ./hosts/kore;
}
```
</details>

<details>
<summary>hosts/kore/default.nix</summary>

```nix
{ config, pkgs, ... }:

{
  imports = [
    ./hardware-configuration.nix
  ];

  networking.hostName = "kore";
  # networking.wireless.enable = true;

  security.sudo.enable = true;

  time.timeZone = "America/Los_Angeles";

  environment.systemPackages = with pkgs; [
    vim
  ];

  nixpkgs.config.allowUnfree = true;

  services.openssh.enable = true;

  system.stateVersion = "19.09"; # Did you read the comment?
}
```
</details>

<details>
<summary>hosts/kore/hardware-configuration.nix</summary>

```nix
{ config, lib, ... }:

let
  pkgs = import ../../../../../vendor/nix/nixpkgs.nix {};
in {
  imports =
    [ "${pkgs.path}/nixos/modules/installer/scan/not-detected.nix"
    ];

  # Use the extlinux boot loader. (NixOS wants to enable GRUB by default)
  boot.loader.grub.enable = false;
  # Enables the generation of /boot/extlinux/extlinux.conf
  boot.loader.generic-extlinux-compatible.enable = true;

  boot.initrd.availableKernelModules = [ ];
  boot.initrd.kernelModules = [ ];
  boot.kernelModules = [ ];
  boot.extraModulePackages = [ ];
  boot.kernelParams = [ "console=ttyS1,115200n8" ];

  nixpkgs.system = "aarch64-linux";

  hardware.enableRedistributableFirmware = true;

  fileSystems."/" =
    { device = "/dev/disk/by-uuid/44444444-4444-4444-8888-888888888888";
      fsType = "ext4";
    };

  swapDevices = [ ];

  nix.maxJobs = lib.mkDefault 4;
}
```

</details>

I'm using morph version 1.2.0.

**Comment by kalbasit** [comment]
I also tried with the nixpkgs given in the example and I got the same issue.

```
  # Pin the deployment package-set to a specific version of nixpkgs
  pkgs = import (builtins.fetchTarball {
    url = "https://github.com/NixOS/nixpkgs-channels/archive/98c1150f2cc62b94b693dce63adc1fbcbfe616f1.tar.gz";
    sha256 = "1mdwn0qrjc8jli8cbi4cfkar6xq15l232r371p4b48v2d4bah3wp";
  }) {};
```

**Comment by johanot** [comment]
@kalbasit Thanks for reporting. Strange... Sounds like the wrong interpreter is being used. Bash? That would explain `command not found`, when the shell tries to exec `use ...`. This could happen of course, if the shebang is not being read/interpreted correctly.

I just found a random switch-to-configuration script from one of my x86_64 systems. The shebang is: `#! /nix/store/c8sbqdfvykyp3lgd2fcphr9rzr3l92vw-perl-5.28.1/bin/perl -I/nix/store/njlwlvhjj5v41wkly5m5fqhl7gm8ipla-perl5.28.1-File-Slurp-9999.26/lib/perl5/site_perl -I/nix/store/3400256pm5dmbnhr1mykvppkgp5mk1c4-perl5.28.1-Net-DBus-1.1.0/lib/perl5/site_perl -I/nix/store/azrfqx8wjmjh5niyghf39nymcz05g077-perl5.28.1-XML-Parser-2.44/lib/perl5/site_perl -I/nix/store/mw5ggn3cqkz7fz18p3n3vbhycgd5id5g-perl5.28.1-XML-Twig-3.52/lib/perl5/site_perl`

Could it be an aarch-variant of: https://lwn.net/Articles/779997/  ? 

Can you paste the shebang of your `switch-to-configuration`-script, and perhaps the kernel version here?

**Comment by kalbasit** [comment]
Yea I see the shebang is perl.

```
#! /nix/store/4h33limcvfdr6ivyjqfbzr6k62y8ipjg-perl-5.28.2/bin/perl -I/nix/store/qb80pgq401g9pc5mjxlmd6ynd76c47xx-perl5.28.2-File-Slurp-9999.27/lib/perl5/site_perl -I/nix/store/xbv2g30ik34g9kixgqylgwjbkn93xkvv-perl5.28.2-Net-DBus-1.1.0/lib/perl5/site_perl -I/nix/store/cwx0j181avl4x8p2cbba3s438a8cxa31-perl5.28.2-XML-Parser-2.44/lib/perl5/site_perl -I/nix/store/r8cqgfn69fz1ckg6ir63d9avc5mgqhpb-perl5.28.2-XML-Twig-3.52/lib/perl5/site_perl
```

Does morph execute `bash switch-to-configuration`? I think it should just execute the `switch-to-configuration` directly given that it's executable with a shebang.

cc @grahamc

**Comment by johanot** [comment]
It looks like the famous kerne bug to be honest :) what is your target kernel version?

No, morph doesn't enforce Bash or any other interpreter.

**Comment by kalbasit** [comment]
@johanot whatever kernel is currently shipped by `nixpkgs-stable` as well as the version used by simple.nix

```
  # Pin the deployment package-set to a specific version of nixpkgs
  pkgs = import (builtins.fetchTarball {
    url = "https://github.com/NixOS/nixpkgs-channels/archive/98c1150f2cc62b94b693dce63adc1fbcbfe616f1.tar.gz";
    sha256 = "1mdwn0qrjc8jli8cbi4cfkar6xq15l232r371p4b48v2d4bah3wp";
  }) {};```

**Comment by ivan** [comment]
Try `uname -a` on the target machine. If it's 4.19.21 or 4.14.99, you can rollback, or run `perl /nix/store/wckw45dkz19ncfi5nzfb736qnhzv77zg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration` to finish the switch-to-configuration step.

**Comment by kalbasit** [comment]
```
[root@kore:~]# uname -a
Linux kore 4.19.61 #1-NixOS SMP Fri Jul 26 07:14:31 UTC 2019 aarch64 GNU/Linux
[root@kore:~]# /nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration
/nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 3: use: command not found
/nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 4: use: command not found
/nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 5: use: command not found
/nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 6: use: command not found
/nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 7: use: command not found
/nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 8: syntax error near unexpected token `('
/nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration: line 8: `use Sys::Syslog qw(:standard :macros);'

[root@kore:~]# head /nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration
#! /nix/store/4h33limcvfdr6ivyjqfbzr6k62y8ipjg-perl-5.28.2/bin/perl -I/nix/store/qb80pgq401g9pc5mjxlmd6ynd76c47xx-perl5.28.2-File-Slurp-9999.27/lib/perl5/site_perl -I/nix/store/xbv2g30ik34g9kixgqylgwjbkn93xkvv-perl5.28.2-Net-DBus-1.1.0/lib/perl5/site_perl -I/nix/store/cwx0j181avl4x8p2cbba3s438a8cxa31-perl5.28.2-XML-Parser-2.44/lib/perl5/site_perl -I/nix/store/r8cqgfn69fz1ckg6ir63d9avc5mgqhpb-perl5.28.2-XML-Twig-3.52/lib/perl5/site_perl

use strict;
use warnings;
use File::Basename;
use File::Slurp;
use Net::DBus;
use Sys::Syslog qw(:standard :macros);
use Cwd 'abs_path';


[root@kore:~]# /nix/store/4h33limcvfdr6ivyjqfbzr6k62y8ipjg-perl-5.28.2/bin/perl -I/nix/store/qb80pgq401g9pc5mjxlmd6ynd76c47xx-perl5.28.2-File-Slurp-9999.27/lib/perl5/site_perl -I/nix/store/xbv2g30ik34g9kixgqylgwjbkn93xkvv-perl5.28.2-Net-DBus-1.1.0/lib/perl5/site_perl -I/nix/store/cwx0j181avl4x8p2cbba3s438a8cxa31-perl5.28.2-XML-Parser-2.44/lib/perl5/site_perl -I/nix/store/r8cqgfn69fz1ckg6ir63d9avc5mgqhpb-perl5.28.2-XML-Twig-3.52/lib/perl5/site_perl /nix/store/80xxmcq4l6ayjb14z3j6mkgdxn6r8kgg-nixos-system-kore-19.09pre-git/bin/switch-to-configuration
-bash: /nix/store/4h33limcvfdr6ivyjqfbzr6k62y8ipjg-perl-5.28.2/bin/perl: cannot execute binary file: Exec format error
[root@kore:~]# nix run nixpkgs.binutils -c readelf -h /nix/store/4h33limcvfdr6ivyjqfbzr6k62y8ipjg-perl-5.28.2/bin/perl
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              EXEC (Executable file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x401270
  Start of program headers:          64 (bytes into file)
  Start of section headers:          15256 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         11
  Size of section headers:           64 (bytes)
  Number of section headers:         28
  Section header string table index: 27
```

I don't think I understand building systems cross-platform.

I thought adding `nixpkgs.system = "aarch64-linux";` to the host configuration is enough to build an aarch64 system from an x86_64 system.

**Comment by johanot** [comment]
@kalbasit Thanks for the extra info. Your kernel is _not_ affected, but the wrong ELF-arch of Perl definitely explains the issue.

Based on your question about `nixpkgs.system`, I now suspect that what you are seeing is another face of #34. Unfortunately most of the `nixpkgs.`-options are not honored by morph, due to the way morph evals machine configs. We are aware of this, and it is clearly a priority to fix it. We just need to decide on the right fix.

**Comment by leotaku** [comment]
@kalbasit If you want a temporary fix before #34 goes away, this is (a simplified version of) what I use in my personal RPI deployment:

```nix
{
  network = {
    description = "RPI3 Server";
    pkgs = (import <nixpkgs> {
      system = "aarch64-linux";
    });
  };

  "nixos-rpi.local" = { config, pkgs, ... }: {
     # stuff...
  };
}
```

**Comment by kalbasit** [comment]
@leotaku thank you, I'll give it a shot then.

**Comment by johanot** [comment]
@kalbasit Is it possible for you to test with: https://github.com/DBCDK/morph/pull/67 ?

**Comment by kalbasit** [comment]
> @kalbasit Is it possible for you to test with: #67 ?

Sure. I'll give it a shot sometime in the next couple of days and get back to you.

**Comment by johanot** [comment]
> Sure. I'll give it a shot sometime in the next couple of days and get back to you.

Perfect. We've merged #67  to master now, since all our tests had good results. We're cutting a release momentarily and PR'ing that against nixpkgs. However; very interested in the result of your test still. If we still see issues, we'll do a bugfix release.

**Comment by adamtulinius** [comment]
Please open a new issue if #67 didn't fix this.
